### PR TITLE
thread::available_parallelism for wasm32-wasip1-threads

### PR DESCRIPTION
**Pull request by surban**
The target has limited POSIX support and provides the `libc::sysconf` function which allows querying the number of available CPUs.

### CODE REVIEW

File: library/std/src/sys/pal/wasi/thread.rs [hunk 1]
```diff
@@ -187,5 +188,15 @@ impl Thread {
 }
 
 pub fn available_parallelism() -> io::Result<NonZero<usize>> {
-    unsupported()
+    cfg_if::cfg_if! {
+        if #[cfg(target_feature = "atomics")] {
+            match unsafe { libc::sysconf(libc::_SC_NPROCESSORS_ONLN) } {
+                -1 => Err(io::Error::last_os_error()),
+                0 => Err(io::Error::UNKNOWN_THREAD_COUNT),
+                cpus => Ok(unsafe { NonZero::new_unchecked(cpus as usize) }),
```
[the remark this answers is not in the document]
Comment (surban) [reply library/std/src/sys/pal/wasi/thread.rs:196]: Implemented.

**Comment by Amanieu** [comment]
While wasi-libc does provide this function, the [current implementation](https://github.com/WebAssembly/wasi-libc/blob/33c208377b272a55afd2cb8607990ef2d3f6c69e/libc-top-half/musl/src/conf/sysconf.c#L262) always returns 1.

@alexcrichton do you have any thoughts on this?

**Comment by surban** [comment]
> While wasi-libc does provide this function, the [current implementation](https://github.com/WebAssembly/wasi-libc/blob/33c208377b272a55afd2cb8607990ef2d3f6c69e/libc-top-half/musl/src/conf/sysconf.c#L262) always returns 1.

Yes, I am aware of this. However, it is possible to link to your customized WASI libc, which interacts with the executor to obtain the real number of processors.

**Comment by Amanieu** [comment]
Fair enough, it doesn't hurt anyways.

@bors r+

**Comment by alexcrichton** [comment]
I agree it's reasonable to hook this up for if/when libc implements this more fully, otherwise it's pretty harmless to have an indirect way to return 1 for now.

**Pull request merged**